File: src/app/model/VerticalSpeedMetric.ts
import { LocationServiceEvent } from './LocationServiceEvent';
import { MetricService } from "../services/metric.service";
import { Metric } from "./Metric";

export class VerticalSpeedMetric extends Metric {

    protected lastLocation: GeolocationPosition | null = null;
    private handler = this.locationEventHandler.bind(this);

    constructor(metricService: MetricService) {
        super('Vertical Speed', 'm/s', metricService);
    }

    startLogging(): void {
        this.metricService.getLocationService().subscribeForLocation(this.handler);
    }

    stopLogging(): void {
        this.metricService.getLocationService().unsubscribeForLocation(this.handler);
    }

    locationEventHandler(event: LocationServiceEvent): void {
        const lastAltitude = this.lastLocation?.coords?.altitude ?? null;
        const currentAltitude = event.location?.coords?.altitude ?? null;
        if (lastAltitude !== null && currentAltitude !== null) {
            const distance = currentAltitude - lastAltitude;
            const time = (event.location.timestamp - this.lastLocation!.timestamp) / 1000.0; // convert to seconds
            const verticalSpeed = distance / time;
            this.addValue(verticalSpeed, new Date(event.location.timestamp));
        } else {
            // If this is the first location event, we can't calculate vertical speed yet
            this.addValue(0, new Date(event.location.timestamp));
        }
        this.lastLocation = event.location;
    }
}

File: android/app/src/main/java/com/openbikecomputer/model/VerticalSpeedMetric.java
package com.openbikecomputer.model;

import com.openbikecomputer.service.IMetricService;
import com.openbikecomputer.service.OnLocationUpdate;

public class VerticalSpeedMetric extends Metric implements OnLocationUpdate {
    protected LocationSample lastLocation = null;

    public VerticalSpeedMetric(IMetricService metricService) {
        super(MetricType.VERTICAL_SPEED, "Vertical Speed", "m/s", metricService);
    }

    @Override
    public void startLogging() {
        metricService.getLocationService().subscribeForLocation(this);
    }

    @Override
    public void stopLogging() {
        metricService.getLocationService().unsubscribeForLocation(this);
    }

    @Override
    public void onLocation(LocationSample sample) {
        Double lastAltitude = lastLocation == null ? null : lastLocation.altitude;
        Double currentAltitude = sample.altitude;
        if (lastAltitude != null && currentAltitude != null) {
            double distance = currentAltitude - lastAltitude;
            double time = (sample.timestamp - lastLocation.timestamp) / 1000.0;
            addValue(distance / time, sample.timestamp);
        } else {
            addValue(0, sample.timestamp);
        }
        lastLocation = sample;
    }
}
